{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore TPM data matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from Bio import SeqIO, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "from override import GENE_NAME_OVERRIDE, GENE_PRODUCT_OVERRIDE\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# feature list indices\n",
    "LEFT_IDX = 0\n",
    "RIGHT_IDX = 1\n",
    "STRAND_IDX = 2\n",
    "LOCUS_IDX = 3\n",
    "GENE_IDX = 4\n",
    "TYPE_IDX = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load TPM data\n",
    "df = pd.read_csv('data/extract_TPM_counts.tsv',sep='\\t').fillna('')\n",
    "\n",
    "# load mapping from sample to condition\n",
    "with open('data/sample2condition.txt','r') as f:\n",
    "    sample2condition = dict(x.strip().split() for x in f.readlines())\n",
    "\n",
    "# load sample to include file\n",
    "samples_provided = False\n",
    "if samples_provided:\n",
    "    with open('config/samples_to_include.txt','r') as f:\n",
    "        samples = list(x.strip() for x in f.readlines())\n",
    "# if none provided, just use all the samples from the sample2condition dict\n",
    "else: \n",
    "    samples = list(sample2condition.keys())\n",
    "    \n",
    "# load the conditions to include file\n",
    "with open('config/conditions_to_include.txt','r') as f:\n",
    "    conditions = list(x.strip() for x in f.readlines())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>locus</th>\n",
       "      <th>start_coord</th>\n",
       "      <th>end_coord</th>\n",
       "      <th>note</th>\n",
       "      <th>translation</th>\n",
       "      <th>gene_len</th>\n",
       "      <th>...</th>\n",
       "      <th>5GB1_pA9_red_tpm</th>\n",
       "      <th>5GB1_pA9_yellow_tpm</th>\n",
       "      <th>5GB1C-5G-La-BR1_tpm</th>\n",
       "      <th>5GB1C-5G-La-BR2_tpm</th>\n",
       "      <th>5GB1C-5G-N-BR1_tpm</th>\n",
       "      <th>5GB1C-5G-N-BR2_tpm</th>\n",
       "      <th>5GB1C-JG15-La-BR1_tpm</th>\n",
       "      <th>5GB1C-JG15-La-BR2_tpm</th>\n",
       "      <th>5GB1C-JG15-N-BR1_tpm</th>\n",
       "      <th>5GB1C-JG15-N-BR2_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EQU24_RS00005</td>\n",
       "      <td>chromosomal replication initiator protein DnaA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>NZ_CP035467.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1317</td>\n",
       "      <td>Derived by automated computational analysis us...</td>\n",
       "      <td>MSALWNNCLAKLENEISSSEFSTWIRPLQAIETDGQIKLLAPNRFV...</td>\n",
       "      <td>1318</td>\n",
       "      <td>...</td>\n",
       "      <td>38.557373</td>\n",
       "      <td>38.810668</td>\n",
       "      <td>37.444214</td>\n",
       "      <td>40.246006</td>\n",
       "      <td>40.100118</td>\n",
       "      <td>33.432274</td>\n",
       "      <td>39.880174</td>\n",
       "      <td>38.355431</td>\n",
       "      <td>30.247582</td>\n",
       "      <td>41.248441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EQU24_RS00010</td>\n",
       "      <td>DNA polymerase III subunit beta</td>\n",
       "      <td>CDS</td>\n",
       "      <td></td>\n",
       "      <td>NZ_CP035467.1</td>\n",
       "      <td>1502</td>\n",
       "      <td>2603</td>\n",
       "      <td>Derived by automated computational analysis us...</td>\n",
       "      <td>MKYIINREQLLVPLQQIVSVIEKRQTMPILSNVLMVFRENTLVMTG...</td>\n",
       "      <td>1102</td>\n",
       "      <td>...</td>\n",
       "      <td>52.552767</td>\n",
       "      <td>52.461746</td>\n",
       "      <td>42.676553</td>\n",
       "      <td>49.210083</td>\n",
       "      <td>46.798476</td>\n",
       "      <td>48.142385</td>\n",
       "      <td>45.465136</td>\n",
       "      <td>46.498139</td>\n",
       "      <td>37.152951</td>\n",
       "      <td>52.902410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EQU24_RS00015</td>\n",
       "      <td>DNA replication/repair protein RecF</td>\n",
       "      <td>CDS</td>\n",
       "      <td>recF</td>\n",
       "      <td>NZ_CP035467.1</td>\n",
       "      <td>3060</td>\n",
       "      <td>4140</td>\n",
       "      <td>Derived by automated computational analysis us...</td>\n",
       "      <td>MSLQKLDIFNVRNIRQASLQPSPGLNLIYGANASGKSSVLEAIFIL...</td>\n",
       "      <td>1081</td>\n",
       "      <td>...</td>\n",
       "      <td>31.350991</td>\n",
       "      <td>34.914128</td>\n",
       "      <td>21.479309</td>\n",
       "      <td>24.204682</td>\n",
       "      <td>22.171104</td>\n",
       "      <td>22.006566</td>\n",
       "      <td>22.658157</td>\n",
       "      <td>22.753325</td>\n",
       "      <td>19.407103</td>\n",
       "      <td>29.834124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EQU24_RS00020</td>\n",
       "      <td>DNA topoisomerase (ATP-hydrolyzing) subunit B</td>\n",
       "      <td>CDS</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>NZ_CP035467.1</td>\n",
       "      <td>4185</td>\n",
       "      <td>6600</td>\n",
       "      <td>Derived by automated computational analysis us...</td>\n",
       "      <td>MSENIKQYDSTNIQVLKGLDAVRKRPGMYIGDTDDGTGLHHMVFEV...</td>\n",
       "      <td>2416</td>\n",
       "      <td>...</td>\n",
       "      <td>74.848501</td>\n",
       "      <td>80.850761</td>\n",
       "      <td>54.959319</td>\n",
       "      <td>64.911376</td>\n",
       "      <td>59.653059</td>\n",
       "      <td>64.648318</td>\n",
       "      <td>69.119079</td>\n",
       "      <td>65.643179</td>\n",
       "      <td>57.590223</td>\n",
       "      <td>68.306759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EQU24_RS00025</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>CDS</td>\n",
       "      <td></td>\n",
       "      <td>NZ_CP035467.1</td>\n",
       "      <td>6825</td>\n",
       "      <td>7062</td>\n",
       "      <td>Derived by automated computational analysis us...</td>\n",
       "      <td>VKTTKYFLTTRMRPDREIIKDEWIQYVVRFPENEHIQFDGRIRRWA...</td>\n",
       "      <td>238</td>\n",
       "      <td>...</td>\n",
       "      <td>50.324948</td>\n",
       "      <td>49.349547</td>\n",
       "      <td>34.539657</td>\n",
       "      <td>36.521074</td>\n",
       "      <td>37.789611</td>\n",
       "      <td>39.358066</td>\n",
       "      <td>38.992158</td>\n",
       "      <td>35.870964</td>\n",
       "      <td>41.462392</td>\n",
       "      <td>40.227192</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 108 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       locus_tag                                         product type  \\\n",
       "0  EQU24_RS00005  chromosomal replication initiator protein DnaA  CDS   \n",
       "1  EQU24_RS00010                 DNA polymerase III subunit beta  CDS   \n",
       "2  EQU24_RS00015             DNA replication/repair protein RecF  CDS   \n",
       "3  EQU24_RS00020   DNA topoisomerase (ATP-hydrolyzing) subunit B  CDS   \n",
       "4  EQU24_RS00025                            hypothetical protein  CDS   \n",
       "\n",
       "  gene_symbol          locus  start_coord  end_coord  \\\n",
       "0        dnaA  NZ_CP035467.1            0       1317   \n",
       "1              NZ_CP035467.1         1502       2603   \n",
       "2        recF  NZ_CP035467.1         3060       4140   \n",
       "3        gyrB  NZ_CP035467.1         4185       6600   \n",
       "4              NZ_CP035467.1         6825       7062   \n",
       "\n",
       "                                                note  \\\n",
       "0  Derived by automated computational analysis us...   \n",
       "1  Derived by automated computational analysis us...   \n",
       "2  Derived by automated computational analysis us...   \n",
       "3  Derived by automated computational analysis us...   \n",
       "4  Derived by automated computational analysis us...   \n",
       "\n",
       "                                         translation  gene_len  ...  \\\n",
       "0  MSALWNNCLAKLENEISSSEFSTWIRPLQAIETDGQIKLLAPNRFV...      1318  ...   \n",
       "1  MKYIINREQLLVPLQQIVSVIEKRQTMPILSNVLMVFRENTLVMTG...      1102  ...   \n",
       "2  MSLQKLDIFNVRNIRQASLQPSPGLNLIYGANASGKSSVLEAIFIL...      1081  ...   \n",
       "3  MSENIKQYDSTNIQVLKGLDAVRKRPGMYIGDTDDGTGLHHMVFEV...      2416  ...   \n",
       "4  VKTTKYFLTTRMRPDREIIKDEWIQYVVRFPENEHIQFDGRIRRWA...       238  ...   \n",
       "\n",
       "   5GB1_pA9_red_tpm  5GB1_pA9_yellow_tpm  5GB1C-5G-La-BR1_tpm  \\\n",
       "0         38.557373            38.810668            37.444214   \n",
       "1         52.552767            52.461746            42.676553   \n",
       "2         31.350991            34.914128            21.479309   \n",
       "3         74.848501            80.850761            54.959319   \n",
       "4         50.324948            49.349547            34.539657   \n",
       "\n",
       "   5GB1C-5G-La-BR2_tpm  5GB1C-5G-N-BR1_tpm  5GB1C-5G-N-BR2_tpm  \\\n",
       "0            40.246006           40.100118           33.432274   \n",
       "1            49.210083           46.798476           48.142385   \n",
       "2            24.204682           22.171104           22.006566   \n",
       "3            64.911376           59.653059           64.648318   \n",
       "4            36.521074           37.789611           39.358066   \n",
       "\n",
       "   5GB1C-JG15-La-BR1_tpm  5GB1C-JG15-La-BR2_tpm  5GB1C-JG15-N-BR1_tpm  \\\n",
       "0              39.880174              38.355431             30.247582   \n",
       "1              45.465136              46.498139             37.152951   \n",
       "2              22.658157              22.753325             19.407103   \n",
       "3              69.119079              65.643179             57.590223   \n",
       "4              38.992158              35.870964             41.462392   \n",
       "\n",
       "   5GB1C-JG15-N-BR2_tpm  \n",
       "0             41.248441  \n",
       "1             52.902410  \n",
       "2             29.834124  \n",
       "3             68.306759  \n",
       "4             40.227192  \n",
       "\n",
       "[5 rows x 108 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for the loaded TPM matrix, which column contains the unique gene ids?\n",
    "LOCUS_ID_COL = 'locus_tag'\n",
    "# use this column to get a full list of all genes for which expression was measured\n",
    "LOCI = list(df[LOCUS_ID_COL].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(98, 4215)\n",
      "(80, 4215)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>locus_tag</th>\n",
       "      <th>sample</th>\n",
       "      <th>EQU24_RS00005</th>\n",
       "      <th>EQU24_RS00010</th>\n",
       "      <th>EQU24_RS00015</th>\n",
       "      <th>EQU24_RS00020</th>\n",
       "      <th>EQU24_RS00025</th>\n",
       "      <th>EQU24_RS00030</th>\n",
       "      <th>EQU24_RS00035</th>\n",
       "      <th>EQU24_RS00040</th>\n",
       "      <th>EQU24_RS00045</th>\n",
       "      <th>...</th>\n",
       "      <th>EQU24_RS22115</th>\n",
       "      <th>EQU24_RS22120</th>\n",
       "      <th>EQU24_RS22125</th>\n",
       "      <th>EQU24_RS22130</th>\n",
       "      <th>EQU24_RS22135</th>\n",
       "      <th>EQU24_RS22140</th>\n",
       "      <th>EQU24_RS22145</th>\n",
       "      <th>EQU24_RS22150</th>\n",
       "      <th>EQU24_RS22155</th>\n",
       "      <th>exp_condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5GB1_FM03_TR1_QC_tpm</td>\n",
       "      <td>48.864921</td>\n",
       "      <td>51.315629</td>\n",
       "      <td>33.906257</td>\n",
       "      <td>64.584281</td>\n",
       "      <td>74.801354</td>\n",
       "      <td>13.090237</td>\n",
       "      <td>50.320911</td>\n",
       "      <td>20.811187</td>\n",
       "      <td>21.910217</td>\n",
       "      <td>...</td>\n",
       "      <td>56.948463</td>\n",
       "      <td>52.002311</td>\n",
       "      <td>60.628466</td>\n",
       "      <td>23.015801</td>\n",
       "      <td>104.488835</td>\n",
       "      <td>379.686212</td>\n",
       "      <td>500.129230</td>\n",
       "      <td>1191.565348</td>\n",
       "      <td>796.964423</td>\n",
       "      <td>uMax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5GB1_FM03_TR2_QC_tpm</td>\n",
       "      <td>52.197450</td>\n",
       "      <td>54.947425</td>\n",
       "      <td>29.979783</td>\n",
       "      <td>72.717824</td>\n",
       "      <td>62.709260</td>\n",
       "      <td>24.874673</td>\n",
       "      <td>71.255094</td>\n",
       "      <td>19.773190</td>\n",
       "      <td>19.421485</td>\n",
       "      <td>...</td>\n",
       "      <td>50.243190</td>\n",
       "      <td>41.376658</td>\n",
       "      <td>66.581832</td>\n",
       "      <td>54.669612</td>\n",
       "      <td>103.864151</td>\n",
       "      <td>401.742811</td>\n",
       "      <td>495.262554</td>\n",
       "      <td>1232.784967</td>\n",
       "      <td>724.291958</td>\n",
       "      <td>uMax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5GB1_FM11_TR1_QC_tpm</td>\n",
       "      <td>25.751902</td>\n",
       "      <td>37.216017</td>\n",
       "      <td>21.716802</td>\n",
       "      <td>53.032824</td>\n",
       "      <td>46.348014</td>\n",
       "      <td>42.426259</td>\n",
       "      <td>58.772307</td>\n",
       "      <td>9.667881</td>\n",
       "      <td>8.212313</td>\n",
       "      <td>...</td>\n",
       "      <td>25.635202</td>\n",
       "      <td>29.059082</td>\n",
       "      <td>59.545627</td>\n",
       "      <td>34.189659</td>\n",
       "      <td>78.264069</td>\n",
       "      <td>208.094267</td>\n",
       "      <td>260.011454</td>\n",
       "      <td>666.589153</td>\n",
       "      <td>461.697526</td>\n",
       "      <td>lowO2_fast_growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5GB1_FM11_TR2_QC_tpm</td>\n",
       "      <td>34.928170</td>\n",
       "      <td>40.094206</td>\n",
       "      <td>27.871558</td>\n",
       "      <td>55.595529</td>\n",
       "      <td>57.992363</td>\n",
       "      <td>26.299890</td>\n",
       "      <td>63.611578</td>\n",
       "      <td>16.724890</td>\n",
       "      <td>11.858533</td>\n",
       "      <td>...</td>\n",
       "      <td>30.765310</td>\n",
       "      <td>37.468337</td>\n",
       "      <td>64.670046</td>\n",
       "      <td>32.369096</td>\n",
       "      <td>59.685923</td>\n",
       "      <td>210.820167</td>\n",
       "      <td>331.355028</td>\n",
       "      <td>860.248081</td>\n",
       "      <td>503.720248</td>\n",
       "      <td>lowO2_fast_growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5GB1_FM12_TR1_tpm</td>\n",
       "      <td>28.322788</td>\n",
       "      <td>27.387700</td>\n",
       "      <td>16.570863</td>\n",
       "      <td>36.795081</td>\n",
       "      <td>27.432559</td>\n",
       "      <td>22.380161</td>\n",
       "      <td>75.688839</td>\n",
       "      <td>17.920465</td>\n",
       "      <td>11.197050</td>\n",
       "      <td>...</td>\n",
       "      <td>16.227307</td>\n",
       "      <td>21.206868</td>\n",
       "      <td>17.075168</td>\n",
       "      <td>18.146243</td>\n",
       "      <td>58.833576</td>\n",
       "      <td>204.760175</td>\n",
       "      <td>300.766500</td>\n",
       "      <td>828.322141</td>\n",
       "      <td>544.969944</td>\n",
       "      <td>lowCH4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4215 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "locus_tag                sample  EQU24_RS00005  EQU24_RS00010  EQU24_RS00015  \\\n",
       "2          5GB1_FM03_TR1_QC_tpm      48.864921      51.315629      33.906257   \n",
       "3          5GB1_FM03_TR2_QC_tpm      52.197450      54.947425      29.979783   \n",
       "4          5GB1_FM11_TR1_QC_tpm      25.751902      37.216017      21.716802   \n",
       "5          5GB1_FM11_TR2_QC_tpm      34.928170      40.094206      27.871558   \n",
       "6             5GB1_FM12_TR1_tpm      28.322788      27.387700      16.570863   \n",
       "\n",
       "locus_tag  EQU24_RS00020  EQU24_RS00025  EQU24_RS00030  EQU24_RS00035  \\\n",
       "2              64.584281      74.801354      13.090237      50.320911   \n",
       "3              72.717824      62.709260      24.874673      71.255094   \n",
       "4              53.032824      46.348014      42.426259      58.772307   \n",
       "5              55.595529      57.992363      26.299890      63.611578   \n",
       "6              36.795081      27.432559      22.380161      75.688839   \n",
       "\n",
       "locus_tag  EQU24_RS00040  EQU24_RS00045  ...  EQU24_RS22115  EQU24_RS22120  \\\n",
       "2              20.811187      21.910217  ...      56.948463      52.002311   \n",
       "3              19.773190      19.421485  ...      50.243190      41.376658   \n",
       "4               9.667881       8.212313  ...      25.635202      29.059082   \n",
       "5              16.724890      11.858533  ...      30.765310      37.468337   \n",
       "6              17.920465      11.197050  ...      16.227307      21.206868   \n",
       "\n",
       "locus_tag  EQU24_RS22125  EQU24_RS22130  EQU24_RS22135  EQU24_RS22140  \\\n",
       "2              60.628466      23.015801     104.488835     379.686212   \n",
       "3              66.581832      54.669612     103.864151     401.742811   \n",
       "4              59.545627      34.189659      78.264069     208.094267   \n",
       "5              64.670046      32.369096      59.685923     210.820167   \n",
       "6              17.075168      18.146243      58.833576     204.760175   \n",
       "\n",
       "locus_tag  EQU24_RS22145  EQU24_RS22150  EQU24_RS22155      exp_condition  \n",
       "2             500.129230    1191.565348     796.964423               uMax  \n",
       "3             495.262554    1232.784967     724.291958               uMax  \n",
       "4             260.011454     666.589153     461.697526  lowO2_fast_growth  \n",
       "5             331.355028     860.248081     503.720248  lowO2_fast_growth  \n",
       "6             300.766500     828.322141     544.969944             lowCH4  \n",
       "\n",
       "[5 rows x 4215 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pivot dataframe to make experiments the rows and genes the columns\n",
    "exp_df = df.set_index('locus_tag')[samples].T.reset_index().rename(columns={'index':'sample'})\n",
    "# add experimental condition column to dataframe\n",
    "exp_df['exp_condition'] = exp_df['sample'].apply(lambda x: sample2condition[x])\n",
    "# filter to only conditions that are requested\n",
    "print(exp_df.shape)\n",
    "exp_df = exp_df[exp_df['exp_condition'].isin(conditions)]\n",
    "print(exp_df.shape)\n",
    "\n",
    "exp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_override_dict(filename):\n",
    "    with open(filename, 'r') as f:\n",
    "        lines = [x.strip().split('\\t') for x in f.readlines()]\n",
    "        override_dict = dict(lines)\n",
    "        \n",
    "    return override_dict\n",
    "\n",
    "gene_ov = load_override_dict(\"data/gene_name_override.txt\")\n",
    "prod_ov = load_override_dict(\"data/gene_product_override.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'EQU24_RS05925': 'likely chaperone for smmo'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prod_ov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_average_tpm_by_condition(df_orig,\n",
    "                                 loci,\n",
    "                                 add_pseudocount=True,\n",
    "                                 pseudocount = 0.01\n",
    "                                ):\n",
    "    '''\n",
    "    Given a matrix of samples by genes, group the samples by their experimental condition.\n",
    "    Then for each gene, calculated the average and standard deviation\n",
    "    '''\n",
    "    \n",
    "    # start with a fresh copy of the dataframe\n",
    "    df = df_orig.copy()\n",
    "    \n",
    "    # add a pseudocount to all genes' TPMs\n",
    "    if add_pseudocount:\n",
    "        df[loci] = df[loci] + pseudocount \n",
    "    \n",
    "    df_means = df[['exp_condition']+loci]\\\n",
    "                    .groupby('exp_condition')\\\n",
    "                    .mean()\\\n",
    "                    .reset_index()\n",
    "    \n",
    "    return df_means\n",
    "    \n",
    "def get_top_n_perc_by_condition(df, loci, n):\n",
    "    '''\n",
    "    Given a dataframe of experimental conditions and the average TPM of each gene,\n",
    "    loop through each condition and rank the genes in descending TPM order. Keep of list\n",
    "    of genes that make the top n% in every condition\n",
    "    '''\n",
    "    num_loci = len(loci)\n",
    "    # keep track of a list of top loci for each condition\n",
    "    # key = exp_condition, value = (gene,tpm)\n",
    "    top_n_loci = {}\n",
    "    \n",
    "    # for each experimental condition row\n",
    "    for i,row in df.iterrows():\n",
    "        exp_cond = row.exp_condition\n",
    "        #print(exp_cond)\n",
    "        # Sort all genes in the row by their tpm counts\n",
    "        tpms = sorted(\n",
    "                list(zip(loci, row[loci].values)),  # zip locus and TPM\n",
    "                key=lambda x: x[1],                # sort by TPM\n",
    "                reverse=True                       # descending TPM order\n",
    "        )\n",
    "    \n",
    "        # keep the top n% of the list\n",
    "        top_n_loci[exp_cond] = tpms[:int(0.01*n*num_loci)]\n",
    "        \n",
    "    # keep only the genes that were in all lists via set intersection\n",
    "    top_n_loci_only = [[y[0] for y in top_n_loci[x]] for x in top_n_loci] # drop tpm\n",
    "    top_n_sets = [set(x) for x in top_n_loci_only]                        # convert to sets\n",
    "    top_n_loci_all_conds = set.intersection(*top_n_sets)  # get intersection of all sets\n",
    "\n",
    "    return top_n_loci_all_conds#, top_n_loci, top_n_sets\n",
    "\n",
    "def get_override_gene(locus_tag,cur_gene):\n",
    "    '''\n",
    "    Given a locus tag, return an overridden gene\n",
    "    '''\n",
    "    return GENE_NAME_OVERRIDE[locus_tag] if locus_tag in GENE_NAME_OVERRIDE else cur_gene\n",
    "\n",
    "\n",
    "def get_features_from_genbank(gb_file,\n",
    "                              override=False):\n",
    "    '''\n",
    "    Given a genbank file, parse out all of it's features into a 5-tuple \n",
    "    of (start_coord, end_coord,locus_tag,gene_symbol,type).\n",
    "    '''\n",
    "    # Use BioPython genbank parser\n",
    "    seq_record = SeqIO.parse(gb_file, \"genbank\").__next__()\n",
    "    \n",
    "    feat_list = []\n",
    "    # Loop over the genome file, get the CDS features on each of the strands\n",
    "    for feature in seq_record.features:\n",
    "        if 'locus_tag' in feature.qualifiers:\n",
    "            # get  locus tag \n",
    "            lt = feature.qualifiers['locus_tag'][0]\n",
    "            # get the gene symbol if available, otherwise leave blank\n",
    "            g = \"\" if 'gene' not in feature.qualifiers else feature.qualifiers['gene'][0]\n",
    "\n",
    "            # if there's a customized override file to rename certain genes, apply that here\n",
    "            if override:\n",
    "                g = get_override_gene(lt,g)\n",
    "            \n",
    "            feat_list.append((feature.location.start.position,\n",
    "                             feature.location.end.position,\n",
    "                             feature.strand,\n",
    "                             lt,\n",
    "                             g,\n",
    "                             feature.type))\n",
    "            \n",
    "    return feat_list\n",
    "    \n",
    "    \n",
    "\n",
    "def flag_potential_operon_loci(pos_feats, neg_feats, min_dist):\n",
    "    '''\n",
    "    Given a all features on the pos and neg strands, flag which ones \n",
    "    may be in an operon at the given min_dist between loci on the same strand.\n",
    "    \n",
    "    Return a set of flagged loci.\n",
    "    '''\n",
    "    \n",
    "    # collect locus tags that are potential operon candidates\n",
    "    operon_candidate_loci = []\n",
    "    \n",
    "    # +-----------------+\n",
    "    # | NEGATIVE STRAND |\n",
    "    # +-----------------+\n",
    "    # if we're in the negative list, check the next feature to the right\n",
    "    for i,(left_coord,right_coord,strand,locus,gene,typee) in enumerate(neg_feats):\n",
    "        # if this isn't the last gene (aka no rightward operon),\n",
    "        # check the min_distance window for other annotations\n",
    "        if i < len(neg_feats) -1:\n",
    "            # get the FOLLOWING feature (because on -1 strand)\n",
    "            upstream_loc = neg_feats[i+1]\n",
    "\n",
    "            # if the left side of the upstream gene is within min_distance, \n",
    "            # collect it\n",
    "            if upstream_loc[LEFT_IDX] < (right_coord + min_dist):\n",
    "                operon_candidate_loci.append(neg_feats[i][LOCUS_IDX])\n",
    "    \n",
    "    # +-----------------+\n",
    "    # | POSITIVE STRAND |\n",
    "    # +-----------------+\n",
    "    # if we're in the positive list, check the next feature to the left\n",
    "    for i,(left_coord,right_coord,strand,locus,gene,typee) in enumerate(pos_feats):\n",
    "        # if this isn't the first gene (aka no leftward operon),\n",
    "        # check the min_distance window for other annotations\n",
    "        if i !=0:\n",
    "            # get the PREVIOUS feature (because on +1 strand)\n",
    "            upstream_loc = pos_feats[i-1]\n",
    "\n",
    "            # if the right side of the upstream gene is within min_distance, \n",
    "            # collect it\n",
    "            if upstream_loc[RIGHT_IDX] > (left_coord - min_dist):\n",
    "                operon_candidate_loci.append(pos_feats[i][LOCUS_IDX])\n",
    "    \n",
    "    return set(operon_candidate_loci)\n",
    "\n",
    "\n",
    "def write_loci_to_file(loci,loci_op_filt, outfile):\n",
    "    '''\n",
    "    Given a list of locus tags, and a list of locus_tags possbily in an operon,\n",
    "    write them to a simple tab delimited file with TRUE next to possible operons\n",
    "    '''\n",
    "    \n",
    "    with open(outfile,'w') as f:\n",
    "        # set header to be \"locus_tag\" \\tab \"op?\" \\newline\n",
    "        f.write('locus_tag\\top?\\n')\n",
    "        \n",
    "        for loc in loci:\n",
    "            op = True if loc in loci_op_filt else False\n",
    "            f.write(f'{loc}\\t{op}\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>locus_tag</th>\n",
       "      <th>exp_condition</th>\n",
       "      <th>EQU24_RS00005</th>\n",
       "      <th>EQU24_RS00010</th>\n",
       "      <th>EQU24_RS00015</th>\n",
       "      <th>EQU24_RS00020</th>\n",
       "      <th>EQU24_RS00025</th>\n",
       "      <th>EQU24_RS00030</th>\n",
       "      <th>EQU24_RS00035</th>\n",
       "      <th>EQU24_RS00040</th>\n",
       "      <th>EQU24_RS00045</th>\n",
       "      <th>...</th>\n",
       "      <th>EQU24_RS22110</th>\n",
       "      <th>EQU24_RS22115</th>\n",
       "      <th>EQU24_RS22120</th>\n",
       "      <th>EQU24_RS22125</th>\n",
       "      <th>EQU24_RS22130</th>\n",
       "      <th>EQU24_RS22135</th>\n",
       "      <th>EQU24_RS22140</th>\n",
       "      <th>EQU24_RS22145</th>\n",
       "      <th>EQU24_RS22150</th>\n",
       "      <th>EQU24_RS22155</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MeOH</td>\n",
       "      <td>23.333155</td>\n",
       "      <td>18.915775</td>\n",
       "      <td>18.453916</td>\n",
       "      <td>18.267805</td>\n",
       "      <td>16.960643</td>\n",
       "      <td>12.377795</td>\n",
       "      <td>43.815536</td>\n",
       "      <td>9.670950</td>\n",
       "      <td>7.302145</td>\n",
       "      <td>...</td>\n",
       "      <td>1298.257682</td>\n",
       "      <td>15.624619</td>\n",
       "      <td>20.208066</td>\n",
       "      <td>26.004364</td>\n",
       "      <td>20.960234</td>\n",
       "      <td>28.719983</td>\n",
       "      <td>93.616437</td>\n",
       "      <td>161.528124</td>\n",
       "      <td>496.990651</td>\n",
       "      <td>280.344047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NO3_lowO2_slow_growth</td>\n",
       "      <td>32.050358</td>\n",
       "      <td>43.656760</td>\n",
       "      <td>21.351623</td>\n",
       "      <td>62.267687</td>\n",
       "      <td>41.684925</td>\n",
       "      <td>31.921455</td>\n",
       "      <td>57.849768</td>\n",
       "      <td>16.885694</td>\n",
       "      <td>14.926147</td>\n",
       "      <td>...</td>\n",
       "      <td>6497.868109</td>\n",
       "      <td>26.273485</td>\n",
       "      <td>28.945133</td>\n",
       "      <td>23.525245</td>\n",
       "      <td>26.432667</td>\n",
       "      <td>35.167264</td>\n",
       "      <td>178.996199</td>\n",
       "      <td>164.083806</td>\n",
       "      <td>433.438735</td>\n",
       "      <td>493.895115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NoCu</td>\n",
       "      <td>44.348687</td>\n",
       "      <td>59.629360</td>\n",
       "      <td>28.268717</td>\n",
       "      <td>56.818319</td>\n",
       "      <td>49.839406</td>\n",
       "      <td>38.394652</td>\n",
       "      <td>81.530362</td>\n",
       "      <td>40.501969</td>\n",
       "      <td>36.576500</td>\n",
       "      <td>...</td>\n",
       "      <td>8345.785345</td>\n",
       "      <td>43.065124</td>\n",
       "      <td>34.380565</td>\n",
       "      <td>44.419579</td>\n",
       "      <td>34.601933</td>\n",
       "      <td>65.339879</td>\n",
       "      <td>253.608495</td>\n",
       "      <td>273.284694</td>\n",
       "      <td>731.052190</td>\n",
       "      <td>1087.621126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NoLanthanum</td>\n",
       "      <td>33.444023</td>\n",
       "      <td>43.689839</td>\n",
       "      <td>23.172675</td>\n",
       "      <td>57.297047</td>\n",
       "      <td>42.367072</td>\n",
       "      <td>41.941657</td>\n",
       "      <td>102.513601</td>\n",
       "      <td>30.226787</td>\n",
       "      <td>19.462312</td>\n",
       "      <td>...</td>\n",
       "      <td>5085.637409</td>\n",
       "      <td>16.423284</td>\n",
       "      <td>35.588138</td>\n",
       "      <td>44.623117</td>\n",
       "      <td>43.201743</td>\n",
       "      <td>21.927260</td>\n",
       "      <td>109.783330</td>\n",
       "      <td>67.277718</td>\n",
       "      <td>211.575175</td>\n",
       "      <td>328.943746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WithLanthanum</td>\n",
       "      <td>35.462185</td>\n",
       "      <td>41.792237</td>\n",
       "      <td>20.644554</td>\n",
       "      <td>57.130166</td>\n",
       "      <td>34.258335</td>\n",
       "      <td>46.201637</td>\n",
       "      <td>110.721781</td>\n",
       "      <td>31.813805</td>\n",
       "      <td>19.438086</td>\n",
       "      <td>...</td>\n",
       "      <td>3942.957792</td>\n",
       "      <td>15.972203</td>\n",
       "      <td>34.318829</td>\n",
       "      <td>49.216725</td>\n",
       "      <td>40.000662</td>\n",
       "      <td>21.220809</td>\n",
       "      <td>98.100610</td>\n",
       "      <td>73.116973</td>\n",
       "      <td>194.389586</td>\n",
       "      <td>319.998959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4214 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "locus_tag          exp_condition  EQU24_RS00005  EQU24_RS00010  EQU24_RS00015  \\\n",
       "0                           MeOH      23.333155      18.915775      18.453916   \n",
       "1          NO3_lowO2_slow_growth      32.050358      43.656760      21.351623   \n",
       "2                           NoCu      44.348687      59.629360      28.268717   \n",
       "3                    NoLanthanum      33.444023      43.689839      23.172675   \n",
       "4                  WithLanthanum      35.462185      41.792237      20.644554   \n",
       "\n",
       "locus_tag  EQU24_RS00020  EQU24_RS00025  EQU24_RS00030  EQU24_RS00035  \\\n",
       "0              18.267805      16.960643      12.377795      43.815536   \n",
       "1              62.267687      41.684925      31.921455      57.849768   \n",
       "2              56.818319      49.839406      38.394652      81.530362   \n",
       "3              57.297047      42.367072      41.941657     102.513601   \n",
       "4              57.130166      34.258335      46.201637     110.721781   \n",
       "\n",
       "locus_tag  EQU24_RS00040  EQU24_RS00045  ...  EQU24_RS22110  EQU24_RS22115  \\\n",
       "0               9.670950       7.302145  ...    1298.257682      15.624619   \n",
       "1              16.885694      14.926147  ...    6497.868109      26.273485   \n",
       "2              40.501969      36.576500  ...    8345.785345      43.065124   \n",
       "3              30.226787      19.462312  ...    5085.637409      16.423284   \n",
       "4              31.813805      19.438086  ...    3942.957792      15.972203   \n",
       "\n",
       "locus_tag  EQU24_RS22120  EQU24_RS22125  EQU24_RS22130  EQU24_RS22135  \\\n",
       "0              20.208066      26.004364      20.960234      28.719983   \n",
       "1              28.945133      23.525245      26.432667      35.167264   \n",
       "2              34.380565      44.419579      34.601933      65.339879   \n",
       "3              35.588138      44.623117      43.201743      21.927260   \n",
       "4              34.318829      49.216725      40.000662      21.220809   \n",
       "\n",
       "locus_tag  EQU24_RS22140  EQU24_RS22145  EQU24_RS22150  EQU24_RS22155  \n",
       "0              93.616437     161.528124     496.990651     280.344047  \n",
       "1             178.996199     164.083806     433.438735     493.895115  \n",
       "2             253.608495     273.284694     731.052190    1087.621126  \n",
       "3             109.783330      67.277718     211.575175     328.943746  \n",
       "4              98.100610      73.116973     194.389586     319.998959  \n",
       "\n",
       "[5 rows x 4214 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_means = get_average_tpm_by_condition(exp_df,LOCI)\n",
    "df_means.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_10_genes = get_top_n_perc_by_condition(df_means, LOCI, 10)\n",
    "top_5_genes = get_top_n_perc_by_condition(df_means, LOCI, 5)\n",
    "top_3_genes = get_top_n_perc_by_condition(df_means, LOCI, 3)\n",
    "top_1_genes = get_top_n_perc_by_condition(df_means, LOCI, 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def maine():\n",
    "    gb_file = \"data/5GB1c_sequence.gb\"\n",
    "    feats = get_features_from_genbank(gb_file,override=True)\n",
    "    feats_filt = [x for x in feats if x[TYPE_IDX] != 'gene']\n",
    "        \n",
    "    # separate pos and neg lists and sort by gene start coordinate\n",
    "    pos_feats = [x for x in feats_filt if x[STRAND_IDX]== 1]\n",
    "    pos_feats.sort(key=lambda x: x[LEFT_IDX])\n",
    "    \n",
    "    neg_feats = [x for x in feats_filt if x[STRAND_IDX]==-1]\n",
    "    neg_feats.sort(key=lambda x: x[RIGHT_IDX])\n",
    "    \n",
    "    # make sure we still keep all features\n",
    "    assert(len(feats_filt) == len(pos_feats)+len(neg_feats))\n",
    "    \n",
    "    print(\"All feats:\", len(feats))\n",
    "    print(\"Filt feats:\", len(feats_filt))\n",
    "    \n",
    "    # flag loci potentially in operons (set of locus ids)\n",
    "    maybe_operon_loci = flag_potential_operon_loci(pos_feats, neg_feats, 120)\n",
    "    \n",
    "    for i in range(1,11):\n",
    "        # get top locus ids in all conditions\n",
    "        top_locs = get_top_n_perc_by_condition(df_means, LOCI, i)\n",
    "        \n",
    "        # filter out loci maybe in operons\n",
    "        top_locs_op_filter_out = [x for x in top_locs if x in maybe_operon_loci]\n",
    "        print(f'Top {i}%: {len(top_locs)} ({len(top_locs) - len(top_locs_op_filter_out)} filtered)')\n",
    "\n",
    "        # save to a two-column txt file\n",
    "        outf = f'output/loci_in_top_{i}perc.txt'\n",
    "        write_loci_to_file(top_locs,top_locs_op_filter_out, outf)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All feats: 8862\n",
      "Filt feats: 4431\n",
      "Top 1%: 14 (8 filtered)\n",
      "Top 2%: 26 (17 filtered)\n",
      "Top 3%: 37 (25 filtered)\n",
      "Top 4%: 56 (35 filtered)\n",
      "Top 5%: 84 (43 filtered)\n",
      "Top 6%: 107 (55 filtered)\n",
      "Top 7%: 127 (65 filtered)\n",
      "Top 8%: 153 (78 filtered)\n",
      "Top 9%: 178 (92 filtered)\n",
      "Top 10%: 197 (103 filtered)\n"
     ]
    }
   ],
   "source": [
    "maine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
